draft-irtf-cfrg-hash-to-curve.md

```diff
@@ -1961,17 +2043,18 @@ Operations:
 2. tv2 = 1 + tv1
 3. tv1 = 1 - tv1
 4. tv3 = inv0(tv1 * tv2)
-5. tv4 = sqrt(-g(Z) * (3 * Z^2 + 4 * A))
+5. tv4 = sqrt(-g(Z) * (3 * Z^2 + 4 * A))    # can be precomputed
```

Comment:
Should we lift this to a constant, then?

Reply:
Hmmm... the problem with lifting this into the "Constants" list above is that it's a different kind of constant---the "Constants" list are parameters, whereas these are values derived from those constants that can be precomputed. I think it's clearer if we make a distinction between these two things, and I don't think the pseudocode suffers so much in the process...

Thoughts?

Reply:
I'm not sure I see the difference. `tv4` depends only on `Z` and `A`, which are both themselves constants. Doesn't that make this the same type of constant? Even if we do consider these different, wouldn't we expect implementations to precompute all parameter and derived constants in practice?

Reply:
Sorry, I was super unclear. What I meant was, the constants in the list (A, B, Z) are choices that you have to make to use this map, no matter how you implement the map.

In contrast, tv4 and tv5 are values you can precompute to speed up evaluating the map, but they're more like implementation details for this specific implementation rather than fundamental parameters. Someone implementing the map in projective coordinates would probably precompute different values, not tv4 and tv5.

Does this make sense?

More generally, I don't think we need to focus too much on precomputation here, because the point isn't to suggest that people follow the details of this implementation, just to make clearer what the high-level costs are (as Thomas says). So it seems like it's OK just to note "can be precomputed," without suggesting that people actually *do* precompute---that's what the constants in the straight-line impl are for.

Reply:
Ah, yes, thanks for clarifying! I forgot that we already have these constants in the straight-line implementation.